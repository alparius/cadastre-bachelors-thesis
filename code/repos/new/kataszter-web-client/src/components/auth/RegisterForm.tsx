import React, { useEffect, useState } from "react";
import { FormattedMessage as FM, useIntl } from "react-intl";
import { NavLink } from "react-router-dom";
import { Form, Grid, Header, Loader, Message } from "semantic-ui-react";

import IRegisterUserData from "../../data/user/RegisterUserData";
import ILabelResponse from "../../requests/response/LabelResponse";
import WrappedRecaptcha from "./WrappedRecaptcha";

interface IRegisterFormProps {
    user: IRegisterUserData;
    response: ILabelResponse | undefined;
    recaptchaToken: string;
    setUser(newUser: IRegisterUserData): void;
    postRegister(recaptchaToken: string): Promise<void>;
    setResponse(resp: ILabelResponse | undefined): void;
    setRecaptchaToken(recaptchaToken: string): void;
}

const RegisterForm: React.FC<IRegisterFormProps> = (props: IRegisterFormProps) => {
    const { user, response, recaptchaToken, setUser, postRegister, setResponse, setRecaptchaToken } = props;
    const { formatMessage: fm } = useIntl();

    const [errorMsgs, setErrorMsgs] = useState(new Array<string>());
    const [fieldErrors, setFieldErrors] = useState(new Array<string>());
    const [showLoader, setShowLoader] = useState(false);
    const [recaptchDOMKey, setRecaptchaDOMKey] = useState(0);

    useEffect(() => {
        if (response && response.label.startsWith("error")) {
            setRecaptchaDOMKey((prevCount) => prevCount + 1);
        }
    }, [response, setRecaptchaToken]);

    const handleRecaptcha = (value: string) => {
        setRecaptchaToken(value);
    };

    const handleChange = (_: any, { name, value }: any) => {
        setUser({ ...user, [name]: value });
    };

    useEffect(() => {
        if (response) {
            setShowLoader(false);
        }
    }, [response]);

    const handleRegisterButton = () => {
        setShowLoader(true);
        setErrorMsgs(new Array<string>());
        setFieldErrors(new Array<string>());
        setResponse(undefined);

        const tobeFieldErrors = new Array<string>();
        const tobeErrorMsgs = new Array<string>();

        Object.entries({ ...user, recaptchaToken: "placeholder" }).forEach((field: [string, string]) => {
            if (field[1] === "") {
                tobeFieldErrors.push(field[0]);
            }
        });
        if (tobeFieldErrors.length !== 0) {
            tobeErrorMsgs.push("formError.emptyFields");
        }

        if (user.password !== user.password2) {
            tobeErrorMsgs.push("formError.passwordMismatch");
            tobeFieldErrors.push("password2");
        }

        if (!user.email.includes("@")) {
            tobeErrorMsgs.push("formError.invalidEmail");
            tobeFieldErrors.push("email");
        }

        if (recaptchaToken === "") {
            tobeErrorMsgs.push("error.recaptcha");
            tobeFieldErrors.push("recaptcha");
        }

        if (tobeFieldErrors.length === 0) {
            postRegister(recaptchaToken);
            setErrorMsgs(new Array<string>());
            setFieldErrors(new Array<string>());
        } else {
            setShowLoader(false);
            setFieldErrors(tobeFieldErrors);
            setErrorMsgs(tobeErrorMsgs);
        }
    };

    return (
        <Grid textAlign="center">
            <Grid.Row>
                <Grid.Column style={{ maxWidth: 450 }}>
                    <Form size="large">
                        {response && (
                            <Message negative>
                                <FM id={response.label} />
                            </Message>
                        )}
                        {errorMsgs.length !== 0 && (
                            <Message negative>
                                {errorMsgs.map((msgID: string) => (
                                    <React.Fragment key={msgID}>
                                        <FM id={msgID} />
                                        <br />
                                    </React.Fragment>
                                ))}
                            </Message>
                        )}
                        <Form.Input
                            error={fieldErrors.includes("name")}
                            name="name"
                            value={user.name}
                            placeholder={fm({ id: "auth.name" })}
                            onChange={handleChange}
                            type="text"
                            fluid
                            icon="address card outline"
                            iconPosition="left"
                        />
                        <Form.Input
                            error={fieldErrors.includes("email")}
                            name="email"
                            value={user.email}
                            placeholder={fm({ id: "auth.email" })}
                            onChange={handleChange}
                            type="text"
                            fluid
                            icon="user outline"
                            iconPosition="left"
                        />
                        <Form.Input
                            error={fieldErrors.includes("password")}
                            name="password"
                            value={user.password}
                            placeholder={fm({ id: "auth.password" })}
                            onChange={handleChange}
                            type="password"
                            fluid
                            icon="lock"
                            iconPosition="left"
                        />
                        <Form.Input
                            error={fieldErrors.includes("password2")}
                            name="password2"
                            value={user.password2}
                            placeholder={fm({ id: "auth.password2" })}
                            onChange={handleChange}
                            type="password"
                            fluid
                            icon="lock"
                            iconPosition="left"
                        />
                        <Form.TextArea
                            error={fieldErrors.includes("message")}
                            name="message"
                            value={user.message}
                            placeholder={fm({ id: "auth.message.please" })}
                            onChange={handleChange}
                            type="text"
                        />

                        <WrappedRecaptcha domkey={recaptchDOMKey} handleRecaptcha={handleRecaptcha} />

                        <Loader active={showLoader} size="large" inline />

                        <Form.Button color="violet" fluid size="large" onClick={handleRegisterButton} style={{ marginTop: "3em" }}>
                            <FM id="auth.registerpage.register" />
                        </Form.Button>
                    </Form>
                </Grid.Column>
            </Grid.Row>

            <Grid.Row style={{ marginTop: "10vh" }}>
                <Grid.Column style={{ maxWidth: 800 }}>
                    <Header as="h3">
                        <FM id="auth.registerpage.loginCall" />{" "}
                        <NavLink to="/login" className="hoverLink">
                            <FM id="auth.registerpage.login" />
                        </NavLink>
                    </Header>
                </Grid.Column>
            </Grid.Row>
        </Grid>
    );
};

export default RegisterForm;
