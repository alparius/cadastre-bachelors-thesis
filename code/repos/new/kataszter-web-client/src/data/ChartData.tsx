interface IChartData {
    key: any;
    data: any;
    counter: number;
}

export default IChartData;
