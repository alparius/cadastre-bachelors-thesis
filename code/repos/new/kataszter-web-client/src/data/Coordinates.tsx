interface ICoordinates {
    lat: number;
    lng: number;
}

export let NewCoordinates: ICoordinates = {
    lat: 0,
    lng: 0
};

export default ICoordinates;
