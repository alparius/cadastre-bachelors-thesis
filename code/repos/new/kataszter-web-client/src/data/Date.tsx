interface IDate {
    year: number;
    month: number;
    day: number;
    comment: string;
}

export let NewDate: IDate = {
    year: 0,
    month: 0,
    day: 0,
    comment: ""
};

export default IDate;
