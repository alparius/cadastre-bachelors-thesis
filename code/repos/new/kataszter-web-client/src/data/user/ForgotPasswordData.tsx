interface IForgotPasswordData {
    email: string;
}

export let NewForgotPasswordData: IForgotPasswordData = {
    email: ""
};

export default IForgotPasswordData;
