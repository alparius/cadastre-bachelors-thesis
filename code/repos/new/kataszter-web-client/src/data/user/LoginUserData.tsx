interface ILoginUserData {
    email: string;
    password: string;
    recaptchaToken: string;
}

export let NewLoginUserData: ILoginUserData = {
    email: "",
    password: "",
    recaptchaToken: ""
};

export default ILoginUserData;
