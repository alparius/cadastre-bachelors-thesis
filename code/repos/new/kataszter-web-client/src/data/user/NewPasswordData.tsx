interface INewPasswordData {
    id: string;
    password: string;
    confirmPassword?: string;
}

export let NewPasswordData: INewPasswordData = {
    id: "",
    password: "",
    confirmPassword: ""
};

export default INewPasswordData;
