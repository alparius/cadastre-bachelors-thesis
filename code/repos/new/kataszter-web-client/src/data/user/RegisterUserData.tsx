interface IRegisterUserData {
    name: string;
    email: string;
    message: string;
    password: string;
    password2?: string;
    recaptchaToken: string;
}

export let NewRegisterUserData: IRegisterUserData = {
    name: "",
    email: "",
    message: "",
    password: "",
    password2: "",
    recaptchaToken: ""
};

export default IRegisterUserData;
