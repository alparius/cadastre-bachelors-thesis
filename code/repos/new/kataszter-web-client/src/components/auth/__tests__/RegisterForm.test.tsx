/*===================
  RegisterForm.unit.test.tsx
 ===================*/
import React from "react";

import { NewRegisterUserData } from "../../../data/user/RegisterUserData";
import { NewTestLabelResponseError, NewTestRegisterUserData } from "../../../util/test/data/test-data";
import { createComponentWithIntlAndRouter, mountWithIntlAndRouter } from "../../../util/test/test-wrapper";
import RegisterForm from "../RegisterForm";

const handlers = {
    setUser: jest.fn(),
    asyncPost: jest.fn()
};

describe("RegisterForm element", () => {
    it("should render the register form with input fields", () => {
        const container = createComponentWithIntlAndRouter(
            <RegisterForm user={NewTestRegisterUserData} response={undefined} {...handlers} />
        ).toJSON();
        expect(container).toMatchSnapshot();
    });

    it("should render a message too when there is a response", () => {
        const container = createComponentWithIntlAndRouter(
            <RegisterForm user={NewTestRegisterUserData} response={NewTestLabelResponseError} {...handlers} />
        ).toJSON();
        expect(container).toMatchSnapshot();
    });

    it("should call the setUser function when the form gets input", () => {
        const container = mountWithIntlAndRouter(<RegisterForm user={NewTestRegisterUserData} response={undefined} {...handlers} />);
        let input = container.find('input[name="name"]');
        input.simulate("change", { target: { value: "test" } });
        input = container.find('input[name="email"]');
        input.simulate("change", { target: { value: "test" } });
        input = container.find('input[name="password"]');
        input.simulate("change", { target: { value: "test" } });
        input = container.find('input[name="password2"]');
        input.simulate("change", { target: { value: "test" } });
        input = container.find('textarea[name="message"]');
        input.simulate("change", { target: { value: "test" } });
        expect(handlers.setUser).toHaveBeenCalledTimes(5);
    });
});
