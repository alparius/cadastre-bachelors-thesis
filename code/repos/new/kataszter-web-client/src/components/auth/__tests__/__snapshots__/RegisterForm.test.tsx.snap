// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterForm element should render a message too when there is a response 1`] = `
<div
  className="ui center aligned grid"
>
  <div
    className="row"
  >
    <div
      className="column"
      style={
        Object {
          "maxWidth": 450,
        }
      }
    >
      <form
        className="ui large form"
        onSubmit={[Function]}
      >
        <div
          className="ui negative message"
        >
          error.test
        </div>
        <div
          className="field"
        >
          <div
            className="ui fluid left icon input"
          >
            <input
              name="name"
              onChange={[Function]}
              placeholder="Full name"
              type="text"
              value="test"
            />
            <i
              aria-hidden="true"
              className="address card outline icon"
              onClick={[Function]}
            />
          </div>
        </div>
        <div
          className="field"
        >
          <div
            className="ui fluid left icon input"
          >
            <input
              name="email"
              onChange={[Function]}
              placeholder="E-mail address"
              type="text"
              value="test"
            />
            <i
              aria-hidden="true"
              className="user outline icon"
              onClick={[Function]}
            />
          </div>
        </div>
        <div
          className="field"
        >
          <div
            className="ui fluid left icon input"
          >
            <input
              name="password"
              onChange={[Function]}
              placeholder="Password"
              type="password"
              value="test"
            />
            <i
              aria-hidden="true"
              className="lock icon"
              onClick={[Function]}
            />
          </div>
        </div>
        <div
          className="field"
        >
          <div
            className="ui fluid left icon input"
          >
            <input
              name="password2"
              onChange={[Function]}
              placeholder="Repeat password"
              type="password"
              value="test"
            />
            <i
              aria-hidden="true"
              className="lock icon"
              onClick={[Function]}
            />
          </div>
        </div>
        <div
          className="field"
        >
          <textarea
            name="message"
            onChange={[Function]}
            onInput={[Function]}
            placeholder="Introduce yourself to the admins of the page. Why would you like to access?"
            rows={3}
            type="text"
            value="test"
          />
        </div>
        <div
          style={
            Object {
              "marginBottom": "3em",
              "marginTop": "1em",
            }
          }
        >
          <div
            className="g-recaptcha"
            data-badge="bottomright"
            data-sitekey="6LfURxEbAAAAAI6TU9GsuydKqmXb3t2TDZwvOYd2"
            data-size="normal"
            data-tabindex="0"
            data-theme="light"
            data-type="image"
            id="g-recaptcha"
          />
        </div>
        <div
          className="ui large inline loader"
        />
        <div
          className="field"
        >
          <button
            className="ui violet large fluid button"
            onClick={[Function]}
            style={
              Object {
                "marginTop": "3em",
              }
            }
          >
            Sign up!
          </button>
        </div>
      </form>
    </div>
  </div>
  <div
    className="row"
    style={
      Object {
        "marginTop": "10vh",
      }
    }
  >
    <div
      className="column"
      style={
        Object {
          "maxWidth": 800,
        }
      }
    >
      <h3
        className="ui header"
      >
        Already have an account?
         
        <a
          aria-current={null}
          className="hoverLink"
          href="/login"
          onClick={[Function]}
        >
          Sign in!
        </a>
      </h3>
    </div>
  </div>
</div>
`;

exports[`RegisterForm element should render the register form with input fields 1`] = `
<div
  className="ui center aligned grid"
>
  <div
    className="row"
  >
    <div
      className="column"
      style={
        Object {
          "maxWidth": 450,
        }
      }
    >
      <form
        className="ui large form"
        onSubmit={[Function]}
      >
        <div
          className="field"
        >
          <div
            className="ui fluid left icon input"
          >
            <input
              name="name"
              onChange={[Function]}
              placeholder="Full name"
              type="text"
              value="test"
            />
            <i
              aria-hidden="true"
              className="address card outline icon"
              onClick={[Function]}
            />
          </div>
        </div>
        <div
          className="field"
        >
          <div
            className="ui fluid left icon input"
          >
            <input
              name="email"
              onChange={[Function]}
              placeholder="E-mail address"
              type="text"
              value="test"
            />
            <i
              aria-hidden="true"
              className="user outline icon"
              onClick={[Function]}
            />
          </div>
        </div>
        <div
          className="field"
        >
          <div
            className="ui fluid left icon input"
          >
            <input
              name="password"
              onChange={[Function]}
              placeholder="Password"
              type="password"
              value="test"
            />
            <i
              aria-hidden="true"
              className="lock icon"
              onClick={[Function]}
            />
          </div>
        </div>
        <div
          className="field"
        >
          <div
            className="ui fluid left icon input"
          >
            <input
              name="password2"
              onChange={[Function]}
              placeholder="Repeat password"
              type="password"
              value="test"
            />
            <i
              aria-hidden="true"
              className="lock icon"
              onClick={[Function]}
            />
          </div>
        </div>
        <div
          className="field"
        >
          <textarea
            name="message"
            onChange={[Function]}
            onInput={[Function]}
            placeholder="Introduce yourself to the admins of the page. Why would you like to access?"
            rows={3}
            type="text"
            value="test"
          />
        </div>
        <div
          style={
            Object {
              "marginBottom": "3em",
              "marginTop": "1em",
            }
          }
        >
          <div
            className="g-recaptcha"
            data-badge="bottomright"
            data-sitekey="6LfURxEbAAAAAI6TU9GsuydKqmXb3t2TDZwvOYd2"
            data-size="normal"
            data-tabindex="0"
            data-theme="light"
            data-type="image"
            id="g-recaptcha"
          />
        </div>
        <div
          className="ui large inline loader"
        />
        <div
          className="field"
        >
          <button
            className="ui violet large fluid button"
            onClick={[Function]}
            style={
              Object {
                "marginTop": "3em",
              }
            }
          >
            Sign up!
          </button>
        </div>
      </form>
    </div>
  </div>
  <div
    className="row"
    style={
      Object {
        "marginTop": "10vh",
      }
    }
  >
    <div
      className="column"
      style={
        Object {
          "maxWidth": 800,
        }
      }
    >
      <h3
        className="ui header"
      >
        Already have an account?
         
        <a
          aria-current={null}
          className="hoverLink"
          href="/login"
          onClick={[Function]}
        >
          Sign in!
        </a>
      </h3>
    </div>
  </div>
</div>
`;
